pandas/_libs/lib.pyx

```diff
@@ -1351,6 +1375,66 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef void get_curr_val(self, curr_seen_t *ret_val):
+        """
+        Gets the current seen val and casts it to ret_val's type.
+
+        While this function does check types to make sure downcasting
+        doesn't happen for the most part, caller is responsible for making
+        sure all int64 <-> uint64 casts are safe
+        """
+        if self.curr_seen is CurrSeen.null_:
+            if curr_seen_t is float64_t or curr_seen_t is complex128_t:
+                ret_val[0] = <curr_seen_t>NaN
+        if self.curr_seen is CurrSeen.bool_:
+            # There's nothing to a bool can be downcast to, it is 1 byte
+            # so no checking
+            ret_val[0] = <curr_seen_t>self.curr_seen_val.bool_
+        elif self.curr_seen is CurrSeen.uint_:
+            if sizeof(ret_val[0]) < sizeof(uint64_t):
+                raise ValueError("Downcasting uint64 is not allowed")
```

Comment:
Do we need to add `except` to the function signature? Want to make sure these don't get swallowed by the cdef

Reply:
I don't think so. I hit these exceptions when I was first writing this up :).